Userland/Libraries/LibGfx/ImageFormats/GIFWriter.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2024, Lucas Chollet <lucas.chollet@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/BitStream.h>
+#include <LibCompress/Lzw.h>
+#include <LibGfx/Bitmap.h>
+#include <LibGfx/ImageFormats/GIFWriter.h>
+
+namespace Gfx {
+
+namespace {
+
+ErrorOr<void> write_header(Stream& stream)
+{
+    // 17. Header
+    TRY(stream.write_until_depleted("GIF87a"sv));
+    return {};
+}
+
+ErrorOr<void> write_logical_descriptor(BigEndianOutputBitStream& stream, Bitmap const& bitmap)
+{
+    // 18. Logical Screen Descriptor
+
+    if (bitmap.width() > NumericLimits<u16>::max() || bitmap.height() > NumericLimits<u16>::max())
+        return Error::from_string_literal("Bitmap size is too big for a GIF");
+
+    TRY(stream.write_value<u16>(bitmap.width()));
+    TRY(stream.write_value<u16>(bitmap.height()));
+
+    // Global Color Table Flag
+    TRY(stream.write_bits(true, 1));
+    // Color Resolution
+    TRY(stream.write_bits(6u, 3));
+    // Sort Flag
+    TRY(stream.write_bits(false, 1));
+    // Size of Global Color Table
+    TRY(stream.write_bits(7u, 3));
+
+    // Background Color Index
+    TRY(stream.write_value<u8>(0));
+
+    // Pixel Aspect Ratio
+    // NOTE: This is not mandatory.
+    TRY(stream.write_value<u8>(0));
+
+    return {};
+}
+
+ErrorOr<void> write_global_color_table(Stream& stream)
+{
+    // 19. Global Color Table
+
+    // FIXME: The color table should include color specific to the image
+    for (u16 i = 0; i < 256; ++i) {
+        TRY(stream.write_value<u8>(i));
+        TRY(stream.write_value<u8>(i));
+        TRY(stream.write_value<u8>(i));
+    }
+    return {};
+}
+
+ErrorOr<void> write_image_data(Stream& stream, Bitmap const& bitmap)
+{
+    // 22. Table Based Image Data
+    auto const pixel_number = static_cast<u32>(bitmap.width() * bitmap.height());
+    auto indexes = TRY(ByteBuffer::create_uninitialized(pixel_number));
+    for (u32 i = 0; i < pixel_number; ++i) {
+        auto const color = Color::from_argb(*(bitmap.begin() + i));
+        if (color.red() != color.green() || color.green() != color.blue() || color.red() != color.blue())
```

Comment:
I might be getting this wrong, but any two of these are enough to fully check the condition, right?

Reply:
You're correct, I removed the last one. Anyway, this won't stay here for long.